crates/compiler/parse/src/state.rs

```diff
@@ -129,6 +129,29 @@ impl<'a> State<'a> {
     pub fn len_region(&self, length: u32) -> Region {
         Region::new(self.pos(), self.pos().bump_column(length))
     }
+
+    pub fn read_file(path: &Path) -> io::Result<Vec<u8>> {
+        use std::io::Read;
+
+        let mut file = File::open(path)?;
+        let file_len = file.metadata()?.len() as usize;
+        let align = 16; // An alignment of 16 is necessary for certain SIMD instructions in parsing.
+        let layout = std::alloc::Layout::from_size_align(file_len, align).unwrap();
+        let buffer_ptr = unsafe { std::alloc::alloc(layout) };
+
+        if buffer_ptr.is_null() {
+            return Err(io::Error::new(io::ErrorKind::Other, "Allocation failed"));
+        }
+
+        // vec now owns the allocated memory and will deallocate it when dropped.
+        // We set the vec's length to be file_len because we're about to read_exact that
+        // many bytes into it.
+        let mut vec = unsafe { Vec::from_raw_parts(buffer_ptr, file_len, file_len) };
```

Comment:
based on https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts, this is UB.

You cannot have a rust `Vec<T>` with an allocation whose alignment is different from the alignment of `T`. When rust drops this vector, it only know the alignment of `T`, which does not match the alignment of the allocation.

also, should we not be using `mmap` to load files into memory?